samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -326,22 +324,22 @@ class TaskInstance(
     * Check each partition assigned to the task is caught to the last offset
     */
   def initCaughtUpMapping() {
-    if (taskContext.getStreamMetadataCache != null) {
+    if (inputStreamMetadata != null && inputStreamMetadata.nonEmpty) {
       systemStreamPartitions.foreach(ssp => {
-        val partitionMetadata = taskContext
-          .getStreamMetadataCache
-          .getSystemStreamMetadata(ssp.getSystemStream, false)
-          .getSystemStreamPartitionMetadata.get(ssp.getPartition)
-
-        val upcomingOffset = partitionMetadata.getUpcomingOffset
-        val startingOffset = offsetManager.getStartingOffset(taskName, ssp)
-          .getOrElse(throw new SamzaException("No offset defined for SystemStreamPartition: %s" format ssp))
-
-        // Mark ssp to be caught up if the starting offset is already the
-        // upcoming offset, meaning the task has consumed all the messages
-        // in this partition before and waiting for the future incoming messages.
-        if(Objects.equals(upcomingOffset, startingOffset)) {
-          ssp2CaughtupMapping(ssp) = true
+        if (inputStreamMetadata.contains(ssp.getSystemStream)) {
```

Comment:
Can there be a scenario where the `inputStreamMetadata` does not contain the `SystemStream` of interest? Would you need to fallback to querying the stream metadata cache from the task context?
If not, it'd be good to at least log a warning in the else block in case we need to troubleshoot in the future.

Reply:
> inputStreamMetadata does not contain the SystemStream of interest

No, that scenario will not happen.  `inputStreamMetadata` map will contain metadata of all input streams in a SamzaContainer(Union of input streams assigned to all the `TaskInstance` in a container). So it will have the stream-metadata for any `SystemStreamPartition` assigned to a `TaskInstance`.